RavenDB-17670 - Queue ETL (Kafka and RabbitMQ)

djordjedjukic opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17670

### Additional description

New feature - Queue ETL - Implementation for Kafka and RabbitMQ.

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
 - Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- It includes Studio work as well

djordjedjukic committed:
RavenDB-17670 Queue ETL  - Kafka implementation

djordjedjukic committed:
RavenDB-17670 Queue ETL:

- client per process
- test connection string
- test sript
- cleanup indexes
- fix transformation script
- modifu tests
- validation

djordjedjukic committed:
RavenDB-17670 rabbitmq wip

ml054 commented on src/Raven.Studio/typescript/configuration.ts:
```diff
@@ -0,0 +1,320 @@
+// This class is autogenerated. Do NOT modify
```
IIRC those files are ignored - please remove them

arekpalinski replied:
done

ml054 commented on src/Raven.Studio/typescript/endpoints.ts:
```diff
@@ -0,0 +1,627 @@
+// This class is autogenerated. Do NOT modify
```
IIRC those files are ignored - please remove them

arekpalinski replied:
done

ayende commented on src/Raven.Client/Documents/Operations/ETL/Queue/QueueConnectionString.cs:
```diff
@@ -0,0 +1,46 @@
+
+    public string Url { get; set; }
+
+    public Dictionary<string, string> KafkaConnectionOptions { get; set; }
```
Does it have to be `Kafka` here? Just `ConnectionOptions`, maybe?

djordjedjukic replied:
It's still in a draft. It's not the final decision. 
We maybe want to have separate connection string classes for each provider.

ayende commented on src/Raven.Server/Commercial/LicenseStatus.cs:
```diff
@@ -170,6 +170,9 @@ public bool CanAutoRenewLetsEncryptCertificate
 
         public bool HasElasticSearchEtl => GetValue<bool>("elasticSearchEtl");
 
+        //public bool HasQueueEtl => GetValue<bool>("queueEtl");
```
This probably needs to have a TODO or something that would make sure that we'll add that later

ayende commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -93,6 +97,7 @@ private void OnProcessRemoved(EtlProcess process)
             List<SqlEtlConfiguration> newSqlDestinations,
             List<OlapEtlConfiguration> newOlapDestinations,
             List<ElasticSearchEtlConfiguration> newElasticSearchDestinations,
+            List<QueueEtlConfiguration> newQueueDestinations,
```
Shouldn't we just pass the DatabaseRecord, instead?

djordjedjukic replied:
We can not use DatabaseRecord, because the ones we passed here are specific to the nodes.

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+    protected override IEnumerator<QueueItem> ConvertTombstonesEnumerator(DocumentsOperationContext context,
+        IEnumerator<Tombstone> tombstones, string collection, bool trackAttachments)
+    {
+        //todo djordje: ignore tombstones
```
Just a reminder about TODO

arekpalinski replied:
done

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+                        Type = string.IsNullOrWhiteSpace(queueItem.Type) ? DefaultType : queueItem.Type,
+                        Source = new Uri(string.IsNullOrWhiteSpace(queueItem.Source) ? DefaultSource : queueItem.Source),
+                        Data = insertItem.TransformationResult
+                    };
```
Let's revise our approach when it comes to Id, Type ans Source. In particular according to the specs id + source should be unique.

1. ID - https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#id - I think that change vector would be suitable

2. Type - https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#type - looks like it's more up to a user's decision. I wonder if we shouldn't provide a way to let him provide it. e.g. `loadToUsers({ ... json message ... }. { type: 'some-type' })`


3. https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#source-1 - also user might be interested in setting it up but as the default I'd say that a node URL + database name + etl task name would be good candidate

djordjedjukic replied:
Discussed directly

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+                        Data = insertItem.TransformationResult
+                    };
+
+                    eventMessage.SetPartitionKey(insertItem.ChangeVector);
```
We already confirmed that it's not related to idempotent kafka producer. So a better choice would be to use source document id - this way for example order and it's orders if put into separate topics would be placed in the same partition what I believe is the demanded behavior.

Also we have to make it configurable to the user because it's very important. See my above comment - we could set an optional parameter of `loadTo()` method where they could set the options to an event - `loadToUsers({ ... json message ... }. { partitionKey: '....' })`

djordjedjukic replied:
Discussed directly

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/BlittableEventBinaryFormatter.cs:
```diff
@@ -0,0 +1,86 @@
+        {
+            foreach (MemoryStream ms in _streams)
+            {
+                _ctx.ReturnMemoryStream(ms);
```
Are we sure that the buffers they hold are now no longer in use at this point?

djordjedjukic replied:
@arekpalinski will take care of it.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/Enumerators/DocumentsToQueueItems.cs:
```diff
@@ -0,0 +1,41 @@
+            if (_docs.MoveNext() == false)
+                return false;
+
+            Current = new QueueItem(_docs.Current, _collection);
```
Can we avoid this allocation?

djordjedjukic replied:
We do the same for all other ETLs.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueDocumentTransformer.cs:
```diff
@@ -0,0 +1,108 @@
+
+        var result = document.TranslateToObject(Context);
+        var index = GetOrAdd(queueName);
+        index.Inserts.Add(new QueueItem(Current) {TransformationResult = result});
```
index?

djordjedjukic replied:
I will fix it.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueDocumentTransformer.cs:
```diff
@@ -0,0 +1,108 @@
+
+    public override void Transform(QueueItem item, EtlStatsScope stats, EtlProcessState state)
+    {
+        if (item.IsDelete == false)
```
How are we going to handle deletes?

djordjedjukic replied:
Currently, we just ignore document tombstones.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueDocumentTransformer.cs:
```diff
@@ -0,0 +1,108 @@
+    {
+        if (_queues.TryGetValue(queueName, out QueueWithEvents queue) == false)
+        {
+            var etlQueue = _config.EtlQueues.Find(x => x.Name.Equals(queueName, StringComparison.OrdinalIgnoreCase));
```
Let's make this a dictionary and avoid `O(N)` lookups

djordjedjukic replied:
_queues are already a dictionary.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+
+    private const string DefaultCloudEventId = "cloud-event-id";
+    private const string DefaultType = "ravendb.etl.put";
+    private const string DefaultSource = "https://ravendb.net/";
```
```suggestion
    private const string DefaultSource = "https://ravendb.net/";
```

Need to come up with a better URI, I think.

djordjedjukic replied:
We will changed it to: Node URL + DB name + Task name.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+    private const string DefaultType = "ravendb.etl.put";
+    private const string DefaultSource = "https://ravendb.net/";
+    public const string QueueEtlTag = "Queue ETL";
+    public string TransactionalId => $"{Database.DbId}-{Name}";
```
I assume we can see those things, let's also include the node tag / url.

djordjedjukic replied:
I will add it.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+    protected override IEnumerator<QueueItem> ConvertTombstonesEnumerator(DocumentsOperationContext context,
+        IEnumerator<Tombstone> tombstones, string collection, bool trackAttachments)
+    {
+        //todo djordje: ignore tombstones
```
Thinking about this, why should we ignore them? Shouldn't we allow the user to send _something_ ?

arekpalinski replied:
I think it could be useful. But not sure what could be _something_. Maybe some dedicated function similar to deletion behavior functions where we just pass the tombstone id and user can just to regular `loadTo...()` call?

I think it's something we can address at later stage once we get initial feedback from the users. I opened the a dedicated issue for that:  https://issues.hibernatingrhinos.com/issue/RavenDB-18653

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+                _kafkaProducer.BeginTransaction();
+
+                foreach (var @event in events)
+                {
```
What control the batch size here?

djordjedjukic replied:
The same mechanism like for other ETLs, in particular, the default batch size is ~8k.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+            }
+            catch (Exception ex)
+            {
+                if (ex is KafkaException) _kafkaProducer.AbortTransaction();
```
Under what scenario do we want to _not_ abort the transaction here?
What happens if the abort fails? It would mask the actual error, no?

djordjedjukic replied:
Agree, I will fix that.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+            catch (Exception ex)
+            {
+                if (ex is KafkaException) _kafkaProducer.AbortTransaction();
+                throw new QueueLoadException(ex.Message, ex);
```
Why not just `throw;` ?

djordjedjukic replied:
Just putting it in a dedicated exception, like for other ETLs.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,212 @@
+        {
+            var commands = simulatedWriter.SimulateExecuteCommandText(record, context);
+
+            summaries.Add(new TopicSummary { TopicName = record.Name, Commands = commands.ToArray() });
```
In here, should we also send to Kafka and then abort? 
IIRC, we do something similar for SQL ETL - to ensure things like schema validation

djordjedjukic replied:
We don't do that in Kafka, because consumers configured with read uncommitted option can consume the message before we abort the transaction.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueHelper.cs:
```diff
@@ -0,0 +1,41 @@
+
+        IProducer<string, byte[]> producer = new ProducerBuilder<string?, byte[]>(config)
+            .SetErrorHandler((producer, error) =>
+            {
+            })
```
Need to at least log events here.

djordjedjukic replied:
Added todo.

ayende commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -136,6 +136,11 @@
     <PackageReference Include="AWSSDK.Glacier" Version="3.7.0.148" />
     <PackageReference Include="AWSSDK.S3" Version="3.7.8.19" />
     <PackageReference Include="Azure.Storage.Blobs" Version="12.11.0" />
+    <PackageReference Include="AWSSDK.Glacier" Version="3.7.0.110" />
+    <PackageReference Include="AWSSDK.S3" Version="3.7.7.5" />
+    <PackageReference Include="Azure.Storage.Blobs" Version="12.10.0" />
```
Was this intentional?

ayende replied:
Note two lines above, we have this with different versions

djordjedjukic replied:
I will revert it.

arekpalinski replied:
Please check it again. It looks we're still having duplicated entries here. Please make sure we add only the relevant packages

ayende commented:
We should also have an option to _delete_ documents once the queue transaction is completed.

djordjedjukic committed:
RavenDB-17670 wip rabbitmq queue

djordjedjukic committed:
RavenDB-17670 fixed issue with wrong url

djordjedjukic committed:
RavenDB-17670 resolve pr comments

djordjedjukic committed:
RavenDB-17670 rework queue settings

arekpalinski committed:
RavenDB-17670 Let's remove the commented licensing related checks. We'll handle it in dedicated issue.

arekpalinski committed:
RavenDB-17670 Using slightly better naming. RabbitMQ connection settings have the whole connection string (including username and password) not just URL

arekpalinski committed:
RavenDB-17670 Renaming QueueProvider to QueueBroker

djordjedjukic committed:
RavenDB-17670 fix for cleanup kafka topic

arekpalinski committed:
RavenDB-17670 Fixing failing Kafka tests

arekpalinski committed:
RavenDB-18652 Workaround for ocassionally corrupted messages. It relates to https://github.com/cloudevents/sdk-csharp/issues/209

arekpalinski committed:
RavenDB-17670 Added ability to pass options to loadTo method in order to let user to provide own values for the headers and set partition key and routing key

arekpalinski committed:
RavenDB-17670 Infrastructure for running tests with the usage of Kafka on CI

arekpalinski commented:
test this please

arekpalinski committed:
RavenDB-17670 Added ability to configure that documents processed by Queue ETL will be deleted

arekpalinski committed:
RavenDB-17670 Let's not extract the tombstones. Currently we don't support doing anything with them. We might want to send something on deletions in the future - it has been reported in dedicated issue,

arekpalinski committed:
RavenDB-17670 Fixing failing tests related to updates in the database record

arekpalinski committed:
RavenDB-17670 Removing unnecessary studio files

arekpalinski committed:
RavenDB-17670 Fixing csproj after rebase

djordjedjukic committed:
RavenDB-17670  complete tests for rabbitmq

Danielle9897 committed:
RavenDB-18586  Kafka & RabbitMQ ETLs

djordjedjukic committed:
RavenDB-17670 expand configuration with exchange type

djordjedjukic committed:
RavenDB-17670 separate declaring and publishing, batching

ml054 commented on src/Raven.Studio/typescript/commands/database/settings/deleteConnectionStringCommand.ts:
```diff
@@ -4,12 +4,20 @@ import endpoints = require("endpoints");
-    constructor(private db: database, private type: Raven.Client.Documents.Operations.ConnectionStrings.ConnectionStringType, private connectionStringName: string) {
+    connectionStringType: Raven.Client.Documents.Operations.ConnectionStrings.ConnectionStringType;
+    
+    constructor(private db: database, private type: StudioEtlType, private connectionStringName: string) {
```
can we use OngoingTask type here?

Danielle9897 replied:
Can't,
`OngoingTaskType` has 'QueueEtl'
The connection string type is `enum ConnectionStringType` with only 'Queue'

Danielle9897 replied:
refactored, as discussed

ml054 commented on src/Raven.Studio/typescript/models/database/dbGroup/databaseGroupGraph.ts:
```diff
@@ -589,7 +594,7 @@ class databaseGroupGraph {
         selection
             .select(".task-desc")
-            .text(x => ongoingTaskModel.mapTaskType(x.type));
+            .text(x => ongoingTaskModel.mapTaskType(x.type as StudioTaskType));
```
todo?

Danielle9897 replied:
removed the 'as' from here

ml054 commented on src/Raven.Studio/typescript/models/database/stats/etlScriptDefinitionCache.ts:
```diff
@@ -12,7 +12,7 @@ class etlScriptDefinitionCache {
     }
 
-    showDefinitionFor(etlType: Raven.Client.Documents.Operations.ETL.EtlType, taskId: number, transformationName: string) {
+    showDefinitionFor(etlType: Raven.Client.Documents.Operations.ETL.EtlType | StudioEtlType, taskId: number, transformationName: string) { // todo - handle etlType 
```
TODO

Danielle9897 replied:
refactored

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskKafkaEtlEditModel.ts:
```diff
@@ -0,0 +1,35 @@
+                    ConnectionStringName: "",
+                    Name: "",
+                },
+            } as Raven.Client.Documents.Operations.OngoingTasks.OngoingTaskQueueEtlDetails);
```
can we avoid as here?

ml054 replied:
or it requires tons of weird stuff to be added?

Danielle9897 replied:
leaving this, as discussed

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskModel.ts:
```diff
@@ -22,8 +23,19 @@ abstract class ongoingTaskModel {
+        return taskListItem.TaskType;
+    }
+    
+    static mapTaskType(taskType: StudioTaskType): TasksNamesInUI {
```
let's remove TaskNamesInUi type - and name this method: formatStudioTaskType - it should return string and be only used in view. We don't want to deal with formated values anywhere else.

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskModel.ts:
```diff
@@ -22,8 +23,19 @@ abstract class ongoingTaskModel {
+    static getStudioTaskType(taskListItem: Raven.Client.Documents.Operations.OngoingTasks.OngoingTask): StudioTaskType {
+        if (taskListItem.TaskType === "QueueEtl") {
+            const task = (taskListItem as Raven.Client.Documents.Operations.OngoingTasks.OngoingTaskQueueEtlListView);
+            return task.BrokerType === "Kafka" ? "KafkaQueueEtl" : "RabbitQueueEtl";
```
let's do switch here to throw when we add new broker type.

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskQueueEtlEditModel.ts:
```diff
@@ -0,0 +1,208 @@
+    dirtyFlag: () => DirtyFlag;
+
+    get studioTaskType(): StudioTaskType {
+        return "KafkaQueueEtl";
```
is it correct?

ml054 replied:
I think you should remove this method - it is misleading.

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskQueueEtlEditModel.ts:
```diff
@@ -0,0 +1,208 @@
+        }
+    }
+
+    toDto(broker: QueueBroker): Raven.Client.Documents.Operations.ETL.Queue.QueueEtlConfiguration { 
```
I think broker should be field of this class. Not external dependency.

ml054 replied:
oh I can see we have 2 subclasses - so let's make this method protected if there will be name clash then this method can be:
protected toDtoInternal(broker: QueueBroker) and then you can have public methods toDto() in child classes.

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskRabbitMqEtlEditModel.ts:
```diff
@@ -0,0 +1,35 @@
+        return "RabbitQueueEtl";
+    }
+    
+    constructor(dto: Raven.Client.Documents.Operations.OngoingTasks.OngoingTaskQueueEtlDetails) {
```
no need for this ctr

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/status/etlScriptDefinitionPreview.ts:
```diff
@@ -7,7 +7,7 @@ class etlScriptDefinitionPreview extends dialogViewModelBase {
     taskName = ko.observable<string>();
     transformationName = ko.observable<string>();
-    etlType = ko.observable<Raven.Client.Documents.Operations.ETL.EtlType>();
+    etlType = ko.observable<Raven.Client.Documents.Operations.ETL.EtlType | StudioEtlType>();  // todo - handle etlType 
```
TODO

Danielle9897 replied:
pls see the comment inside this file
as I want to show the type in the preview if we know it

    // if we come here from tasks-list-view we know studio-etl-type
    // if we come from stats-view we only know server-etl-type

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/status/etlScriptDefinitionPreview.ts:
```diff
@@ -17,12 +17,13 @@ class etlScriptDefinitionPreview extends dialogViewModelBase {
     };
     
-    constructor(etlType: Raven.Client.Documents.Operations.ETL.EtlType, 
+    constructor(etlType: Raven.Client.Documents.Operations.ETL.EtlType | StudioEtlType, // todo - handle etlType 
```
TODO

Danielle9897 replied:
same - see comment in side this file

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/status/ongoingTasksStats.ts:
```diff
@@ -59,7 +59,7 @@ type trackItemContext = {
     transformationName: string;
     taskId: number;
-    etlType: Raven.Client.Documents.Operations.ETL.EtlType;
+    etlType: Raven.Client.Documents.Operations.ETL.EtlType | StudioEtlType; // todo - should be handled in the future for ongoing stats graph view
```
TODO

Danielle9897 replied:
here I returned it to be just `EtlType`

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/ongoingTasks.ts:
```diff
@@ -550,18 +600,28 @@ class ongoingTasks extends viewModelBase {
     }
+    
+    // todo refactor.. shorter..
+    getTaskNameForUI(taskType: StudioTaskType) {
```
?

Danielle9897 replied:
done - used the shorter syntax

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/transformationScriptSyntax.ts:
```diff
@@ -49,6 +49,8 @@ class transformationScriptSyntax extends dialogViewModelBase {
                         break;
                 }
                 break;
+            case "Queue": // TODO
```
we need a way to distinguish between queue types here

ml054 commented on src/Raven.Studio/typings/_studio/dto.d.ts:
```diff
@@ -720,11 +720,12 @@ interface scrollColorConfig {
 
 type etlScriptDefinitionCacheItem = {
-    etlType: Raven.Client.Documents.Operations.ETL.EtlType;
+    etlType: Raven.Client.Documents.Operations.ETL.EtlType | StudioEtlType; // todo - handle etlType
```
TODO

Danielle9897 replied:
here I returned it to be just `EtlType`

ml054 commented on src/Raven.Studio/typings/_studio/dto.d.ts:
```diff
@@ -820,7 +821,13 @@ interface TimeSeriesOperation extends Raven.Client.Documents.Operations.TimeSeri
 }
 
-type TasksNamesInUI = "External Replication" | "RavenDB ETL" | "SQL ETL" | "OLAP ETL" | "Backup" | "Subscription" | "Replication Hub" | "Replication Sink" | "Elasticsearch ETL";
+type TasksNamesInUI = "External Replication" |  "Replication Hub" | "Replication Sink" | "Backup" | "Subscription" | 
```
remove this type

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editKafkaEtlTask.html:
```diff
@@ -0,0 +1,295 @@
+                    </div>
+                    <div>
+                        <div data-bind="if: createNewConnectionString">
+<!--                            <div class="panel-addon" data-bind="with: testConnectionResult">-->
+<!--                                <div class="padding bg-success small" data-bind="visible: Success">-->
+<!--                                    <div>Successfully connected to the node</div>-->
+<!--                                </div>-->
+<!--                                <div data-bind="if: !Success">-->
+<!--                                    <div class="padding bg-danger small">-->
+<!--                                        <h2>Connection test failed!</h2>-->
+<!--                                        <span data-bind="text: $root.fullErrorDetailsVisible() ? Error : $root.shortErrorText()"></span>-->
+<!--                                        <div>-->
+<!--                                            <a href="#" data-bind="click: $root.fullErrorDetailsVisible.toggle.bind($root.fullErrorDetailsVisible), text: $root.fullErrorDetailsVisible() ? 'hide details' : 'show details'"></a>-->
+<!--                                        </div>-->
+<!--                                    </div>-->
+<!--                                    <div class="padding bg-info small">-->
```
remove?

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/editRabbitMqEtlTask.html:
```diff
@@ -0,0 +1,294 @@
+                    </div>
+                    <div>
+                        <div data-bind="if: createNewConnectionString">
+<!--                            <div class="panel-addon" data-bind="with: testConnectionResult">-->
+<!--                                <div class="padding bg-success small" data-bind="visible: Success">-->
+<!--                                    <div>Successfully connected to the node</div>-->
+<!--                                </div>-->
+<!--                                <div data-bind="if: !Success">-->
+<!--                                    <div class="padding bg-danger small">-->
+<!--                                        <h2>Connection test failed!</h2>-->
+<!--                                        <span data-bind="text: $root.fullErrorDetailsVisible() ? Error : $root.shortErrorText()"></span>-->
+<!--                                        <div>-->
+<!--                                            <a href="#" data-bind="click: $root.fullErrorDetailsVisible.toggle.bind($root.fullErrorDetailsVisible), text: $root.fullErrorDetailsVisible() ? 'hide details' : 'show details'"></a>-->
+<!--                                        </div>-->
+<!--                                    </div>-->
+<!--                                    <div class="padding bg-info small">-->
+<!--                                        <h2>About this error</h2>-->
+<!--                                        <div>Each RavenDB server has both HTTP and TCP endpoints. While the first one is used for system management and client-server rest request, the second is used for inter-server and advanced client-server communications.</div>-->
+<!--                                        <div>The connection tests the TCP endpoint only after a successful HTTP connection.</div>-->
+<!--                                        <div data-bind="visible: HTTPSuccess">It appears that the current server was able to connect to the desired server through HTTP, but failed connecting to it using TCP.</div>-->
+<!--                                        <div data-bind="visible: !HTTPSuccess">It appears that the current server could not connect to the desired node through HTTP.</div>-->
+<!--                                        <div>-->
```
remove?

Danielle9897 replied:
removed

ml054 commented on src/Raven.Studio/wwwroot/App/views/database/tasks/transformationScriptSyntax.html:
```diff
@@ -40,6 +40,12 @@ <h3 class="margin-bottom margin-bottom-lg margin-top">Sample OLAP ETL Transforma
+           <div data-bind="if: etlType() === 'Queue'" class="copy-area">
+               <h3 class="margin-bottom margin-bottom-lg margin-top">Sample Queue ETL Transformation Script</h3>
+               <div class="margin-bottom">
+                   TODO...
```
TODO

djordjedjukic committed:
RavenDB-17670 rework test script

djordjedjukic committed:
RavenDB-17670 rework declaring queues and exchanges

arekpalinski committed:
RavenDB-17670 Setting up ETL to RabbitMQ tests for CI

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueHelper.cs:
```diff
@@ -0,0 +1,52 @@
+    public static IModel CreateRabbitMqClient(QueueConnectionString connectionString, string transactionalId)
+    {
+        var connectionFactory = new ConnectionFactory() { Uri = new Uri(connectionString.RabbitMqConnectionSettings.ConnectionString) };
+        var connection = connectionFactory.CreateConnection();
```
This returns disposable object that we need to dispose as well

djordjedjukic replied:
It's  disposed in QueueEtl class, method Dispose.
Line 394. 
@arekpalinski Do we need to do that somewhere else?

arekpalinski replied:
We dispose there the object returned by `connection.CreateModel();` that we have below, don't we? I think we still need to dispose `connection`

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueHelper.cs:
```diff
@@ -0,0 +1,52 @@
+            config.Set(option.Key, option.Value);
+        }
+
+        IProducer<string, byte[]> producer = new ProducerBuilder<string?, byte[]>(config)
```
Please fix the warning

djordjedjukic replied:
I will fix it.

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,417 @@
+
+        SetUpExchangesAndQueues(messages);
+        
+        foreach (var message in messages)
```
`messages` name is confusing. When I looked at the code first time I was wondering why we open a transaction per message

djordjedjukic replied:
Ok, I will find better names

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Queue/QueueEtl.cs:
```diff
@@ -0,0 +1,417 @@
+        foreach (var record in records)
+        {
+            var messages = simulatedWriter.SimulateExecuteMessages(record, context);
+            // todo: message should have body (content) and options
```
I think it's done already

djordjedjukic replied:
It is, I will remove it.

djordjedjukic committed:
RavenDB-17670 remove comments

Danielle9897 committed:
RavenDB-18586 Fix review comments

djordjedjukic committed:
RavenDB-17670 map basic properties, code cleanup

djordjedjukic committed:
RavenDB-17670 delete queue tasks

Danielle9897 committed:
RavenDB-18586 Test Kafka server connection (+ fix for ElasticSearch)

ml054 committed:
RavenDB-17670 etl subtype in stats

ml054 committed:
RavenDB-17670 etl subtype in stats - studio part

arekpalinski committed:
RavenDB-18786 Handle import and export of queue etls (kafka and rabbitmq)

djordjedjukic committed:
RavenDB-17670 rework test connection string handler

djordjedjukic committed:
RavenDB-17670 cleanup csproj file

djordjedjukic committed:
RavenDB-17670 add headers tests

djordjedjukic committed:
RavenDB-17670 rework rabbitmq IModel and IConnection

Danielle9897 committed:
RavenDB-18586 Fix review comments (Using StudioEtlType in stats view)

ml054 committed:
RavenDB-17670 url -> bootstrap servers

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskModel.ts:
```diff
@@ -69,22 +53,67 @@ abstract class ongoingTaskModel {
+                switch (serverEtlSubType) {
+                    case "Kafka": return "Kafka";
+                    case "RabbitMq": return "RabbitMQ";
+                    default: genUtils.assertUnreachable(serverEtlSubType as never, "Unknown serverEtlSubType: " + serverEtlType);
```
why you force as never here?

ml054 replied:
oh because it is string and it will fail in runtime. ok this is fine.

Danielle9897 committed:
RavenDB-18586 Studio: use BootstrapServers

djordjedjukic committed:
RavenDB-17670 resolve PR comments

djordjedjukic committed:
RavenDB-17670 resolve PR comments

djordjedjukic committed:
RavenDB-17670 resolve PR issues

arekpalinski committed:
RavenDB-17670 Fixing warnings and references in csproj

arekpalinski committed:
RavenDB-17670 Adjusting how we sent the connection options in order to test connection to Kafka and RabbitMQ

arekpalinski committed:
RavenDB-17670 Fixing build

arekpalinski committed:
RavenDB-17670 No need to send the provided options back to the studio

Danielle9897 committed:
RavenDB-18586 Test Connection work for both Kafka & RabbitMQ + Kafka BootstrapServers validation

Danielle9897 committed:
RavenDB-18586 Add Queue tasks to Export & Import

arekpalinski committed:
RavenDB-18786 Fixing export of queue etls (BrokerType property was missing)

arekpalinski committed:
RavenDB-17670 Providing certificate when we are testing connection to Kafka with the usage of Raven cert. Some renames of classes.

arekpalinski committed:
RavenDB-17670  Fixing stop batch conditions

arekpalinski committed:
RavenDB-17670 Improved logging of Kafka producer

djordjedjukic committed:
RavenDB-17670 expand GetOngoingTaskInfoOperation with elasticsearch

arekpalinski committed:
RavenDB-17670 When a process is stopped then we need to dispose the kafka producer and connection to rabbitmq

arekpalinski committed:
RavenDB-17670 Disposing Kafka admin client

ml054 committed:
RavenDB-18586 Studio: Queue ETL - fixed typings

ppekrol commented on src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs:
```diff
@@ -60,5 +60,7 @@ public enum DatabaseRecordItemType
         ElasticSearchEtls = 1 << 22,
         PostgreSQLIntegration = 1 << 23,
+        QueueConnectionStrings = 1 << 24,
+        QueueEtls = 1 << 24,
```
should be 25?

arekpalinski replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Queue/QueueBroker.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Raven.Client.Documents.Operations.ETL.Queue;
+
+public enum QueueBroker
```
QueueBrokerType

arekpalinski replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Queue/QueueConnectionString.cs:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    public string GetUrl()
```
internal?

arekpalinski replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/Queue/Handlers/QueueEtlConnectionHandler.cs:
```diff
@@ -0,0 +1,102 @@
+{
+    public class QueueEtlConnectionHandler : DatabaseRequestHandler
+    {
+        [RavenAction("/databases/*/admin/etl/queue/test-connection/kafka", "POST", AuthorizationStatus.DatabaseAdmin)]
```
should't this be the other way around?
/databases/*/admin/etl/queue/test-connection/kafka => /databases/*/admin/etl/queue/kafka/test-connection

arekpalinski replied:
done

ppekrol commented on src/Raven.Server/ServerWide/JsonDeserializationCluster.cs:
```diff
@@ -106,6 +107,8 @@ public class JsonDeserializationCluster : JsonDeserializationBase
 
         public static Func<BlittableJsonReaderObject, ElasticSearchEtlConfiguration> ElasticSearchEtlConfiguration = GenerateJsonDeserializationRoutine<ElasticSearchEtlConfiguration>();
 
+        public static Func<BlittableJsonReaderObject, QueueEtlConfiguration> QueueEtlConfiguration = GenerateJsonDeserializationRoutine<QueueEtlConfiguration>();
```
readonly

arekpalinski replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/Queue/Handlers/QueueEtlConnectionHandler.cs:
```diff
@@ -0,0 +1,102 @@
+            }
+        }
+
+        [RavenAction("/databases/*/admin/etl/queue/test-connection/rabbitmq", "POST", AuthorizationStatus.DatabaseAdmin)]
```
same here?

arekpalinski replied:
done

ml054 committed:
RavenDB-18586 fixed SSL issues

Danielle9897 committed:
RavenDB-18586 Added info when 'Using RavenDB Cert' is ON (in Kafka connection string)

Danielle9897 committed:
RavenDB-18586 added sample syntax

Danielle9897 committed:
RavenDB-18586 support Task DestinationType

Danielle9897 committed:
RavenDB-18586 kafkaServerUrl => bootstrapServers fixes

Danielle9897 committed:
RavenDB-18586 Add Kafka & RabbitMQ info to cluster-dashboard

ml054 committed:
RavenDB-17670 UI cleanup

arekpalinski committed:
RavenDB-17670 Fixing PR comments

Danielle9897 committed:
RavenDB-18586 Fix compilation + Add to About page + Add options to sample syntax

arekpalinski committed:
RavenDB-17670 Adjustments related to Rabbit MQ implementation:
- removing ExchangeType as an option (instead we'll default to Direct type when we automatically define the queues and exchanges)
- added option SkipAutomaticQueueDeclaration that will allow users to force ETL process to not deal with his config at all
- added validation of property names for options instance

Danielle9897 committed:
RavenDB-18586 Add test script area

Danielle9897 committed:
RavenDB-18586 Add toggle for RabbitMQ - Skip the Automatic Declaration

Danielle9897 committed:
RavenDB-18586 Add load options to test results area
